Add inline definitions (and simplify a bit) some accessors.

sammy-tri opened the pull request:
In a particular simulation (not in drake master), this results in up
to 10% speedup improvement.

I mostly only bothered with the accessors which were visibile in
callgrind output.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/8847)
<!-- Reviewable:end -->

sammy-tri commented:
+@jwnimmer-tri for all review per platform rotation schedule.  (let me know if you think a feature review is required)

---

Review status: 0 of 8 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8847#-:-LD3fk064xe27qyGhATI:b-6enzxh)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> all.

FYI I was curious why these tweaks weren't already done in the original MABLAB codebase (calling this C++ via mex), since that was relatively well-profiled.  The answer is that these all used to be public member fields, but were later (via TRI) refactored to use getters, but the getters did not use inlining.  Oops.

---

Reviewed 8 of 8 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[multibody/rigid_body.h, line 122 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8847#-LD3kDEn4_V-XVCxKrhL:-LD3kDEn4_V-XVCxKrhM:bm8qn3z) ([raw file](https://github.com/robotlocomotion/drake/blob/ee5acbe914cab1c98aeb767bc7d83d20eb81731d/multibody/rigid_body.h#L122)):*
> ```objc
>     DRAKE_DEMAND(joint_ != nullptr);
>     return (*joint_.get());
>   }
> ```

While the git history of the exception text (from cc file) is ... exciting, I see no reason to throw that diagnostic message away?

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8847#-:-LD3jP-HBl5M_crID59v:b-75v0al)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 3 files at r2.
Review status: 8 of 10 files reviewed at latest revision, 1 unresolved discussion.

---

*[multibody/rigid_body.h, line 122 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8847#-LD3kDEn4_V-XVCxKrhL:-LD3qPf2971xeaDITqCn:b-h70fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ee5acbe914cab1c98aeb767bc7d83d20eb81731d/multibody/rigid_body.h#L122)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

While the git history of the exception text (from cc file) is ... exciting, I see no reason to throw that diagnostic message away?
</blockquote></details>

Oh, especially if there were unit tests covering this -- why remove the throw?

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8847)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Missed tests which checked that RigidBodyTree::compile() fails if a body is missing a joint.  Deleted one and moved the check that the other was looking for.

---

Review status: 7 of 10 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8847#-:-LD3ppyzA3-MYmm1beXI:b-b54rk)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Review status: 7 of 10 files reviewed at latest revision, 1 unresolved discussion.

---

*[multibody/rigid_body.h, line 122 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8847#-LD3kDEn4_V-XVCxKrhL:-LD3s6yn4YP4yUOVoSTw:b77chch) ([raw file](https://github.com/robotlocomotion/drake/blob/ee5acbe914cab1c98aeb767bc7d83d20eb81731d/multibody/rigid_body.h#L122)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Oh, especially if there were unit tests covering this -- why remove the throw?
</blockquote></details>

The throw() was causing the function not to be inlined even when it was declared in the header.  According to callgrind, this function was actually the worst offender in the death-by-10-million-cuts this PR addresses, so I thought it was worth losing the descriptive exception.

I moved the check the unit tests were looking for into RBT::compile().

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8847)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 3 files at r2, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[multibody/rigid_body.h, line 121 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8847#-LD3ugkNC38K6dIW7auB:-LD3ugkNC38K6dIW7auC:b-z7rxxl) ([raw file](https://github.com/robotlocomotion/drake/blob/3fa407d7352847e28e289bd7173283b3b47f826c/multibody/rigid_body.h#L121)):*
> ```objc
>   const DrakeJoint& getJoint() const {
>     DRAKE_DEMAND(joint_ != nullptr);
>     return (*joint_.get());
> ```

BTW Don't use `()`s on `return` -- its not a function.  (Yes, this is from the pre-existing code, but its reworked a little here anyway, might as well fix the nit?)

FYI Also the `get()` isn't needed -- `return *joint_;` should work.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8847)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Review status: 8 of 10 files reviewed at latest revision, 1 unresolved discussion.

---

*[multibody/rigid_body.h, line 121 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8847#-LD3ugkNC38K6dIW7auB:-LD772_B79SGG1UGkwcM:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/3fa407d7352847e28e289bd7173283b3b47f826c/multibody/rigid_body.h#L121)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW Don't use `()`s on `return` -- its not a function.  (Yes, this is from the pre-existing code, but its reworked a little here anyway, might as well fix the nit?)

FYI Also the `get()` isn't needed -- `return *joint_;` should work.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8847)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8847)*
<!-- Sent from Reviewable.io -->
